sklearn/tree/_criterion.pyx

```diff
@@ -346,50 +397,165 @@ cdef class ClassificationCriterion(Criterion):
+            offset += self.sum_stride
+
+        # Compute sum_missing
+        for p in range(end-n_missing, end):
+            i = samples[p]
+
+            # w is originally set to be 1.0, meaning that if no sample weights
+            # are given, the default weight of each sample is 1.0
+            if sample_weight != NULL:
+                w = sample_weight[i]
+
+            # Count weighted class frequency for each target
+            offset = i * y_stride
+
+            for k in range(self.n_outputs):
+                c = <SIZE_t> y[offset + k]
+                sum_missing[k * self.sum_stride + c] += w
+            self.weighted_n_node_missing += w
+
+        # Using the computed sum_missing, separate the sum_available
+        # from the sum_total
+        # {sum_total} --> {sum_available} + {sum_missing}
+        for k in range(self.n_outputs):
+            offset = k * self.sum_stride
+            for c in range(n_classes[k]):
+                i = c + offset
+                sum_available[i] = sum_total[i] - sum_missing[i]
+        self.weighted_n_node_available = self.weighted_n_node_samples - self.weighted_n_node_missing
+
+    cdef void reset(self, bint include_missing=True) nogil:
+        """Reset the criterion at pos=start. 
+
+        Move the missing values also to the right, if include_missing is True.
+        """
         cdef double* sum_total = self.sum_total
         cdef double* sum_left = self.sum_left
         cdef double* sum_right = self.sum_right
+        cdef double* sum_missing = self.sum_missing
+        cdef double* sum_available = self.sum_available
 
         cdef SIZE_t* n_classes = self.n_classes
         cdef SIZE_t k
 
-        for k in range(self.n_outputs):
-            memset(sum_left, 0, n_classes[k] * sizeof(double))
-            memcpy(sum_right, sum_total, n_classes[k] * sizeof(double))
+        self.pos = self.start
 
-            sum_total += self.sum_stride
-            sum_left += self.sum_stride
-            sum_right += self.sum_stride
+        # If missing values are to be sent to right or already in the right
+        if include_missing or (self.missing_direction != MISSING_DIR_LEFT):
+            self.weighted_n_left = 0.0
+            self.weighted_n_right = self.weighted_n_node_samples
 
-    cdef void reverse_reset(self) nogil:
-        """Reset the criterion at pos=end."""
-        self.pos = self.end
+            for k in range(self.n_outputs):
+                # Move everything from left to the right partition
+                memset(sum_left, 0, n_classes[k] * sizeof(double))
+                memcpy(sum_right, sum_total, n_classes[k] * sizeof(double))
 
-        self.weighted_n_left = self.weighted_n_node_samples
-        self.weighted_n_right = 0.0
+                sum_total += self.sum_stride
+                sum_left += self.sum_stride
+                sum_right += self.sum_stride
+
+        # If the missing values are not to be moved from left to right
+        else:
```

Comment:
Do we need this `else` clause? Couldnt we reset always by moving the missing values to the right? Then, we necessary, we could call `move_missing_left`. This would avoid some of this logic to be repeated.